chalk-derive/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use quote::{format_ident, quote, ToTokens};
+use syn::{parse_macro_input, Data, DeriveInput, GenericParam, Ident, TypeParamBound};
+
+#[proc_macro_derive(Fold, attributes(has_type_family))]
+pub fn derive_fold(item: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(item as DeriveInput);
+
+    let name = input.ident;
+    let (_, ty_generics, where_clause_ref) = input.generics.split_for_impl();
+    let body = derive_fold_body(input.data);
+
+    // Allow custom Fold parameter using the `#[has_type_family(Param)]` attribute
+    let mut generics = if let Some(attr) = input
+        .attrs
+        .iter()
+        .find(|a| a.path.is_ident("has_type_family"))
+    {
+        let arg = attr
+            .parse_args::<proc_macro2::TokenStream>()
+            .expect("Expected has_type_family argument");
+        quote! { < #arg > } // TODO extend instead of overriding?
+    } else {
+        ty_generics.to_token_stream()
```

Comment:
OK so we assume that one of the following is true:

* there is a `#[has_type_family]` attribute
* there is a parameter `G: HasTypeFamily`
* there is a single generic parameter and it is a type family

Right?

Reply:
The third option should also work when there's more parameters, I think